py-polars/polars/io/_utils.py

```diff
@@ -193,3 +193,77 @@ def _process_http_file(path: str, encoding: str | None = None) -> BytesIO:
             return BytesIO(f.read())
         else:
             return BytesIO(f.read().decode(encoding).encode("utf8"))
+
+
+def _prepare_write_file_arg(
+    file: str | Path | BytesIO,
+    encoding: str | None = None,
+    **kwargs: Any,
+) -> ContextManager[str | BytesIO]:
+    """
+    Prepare file argument.
+
+    Utility for write_[parquet]. (not to be used by scan_[parquet]).
+    Returned value is always usable as a context.
+
+    A :class:`BytesIO` file is returned as a :class:`BytesIO`.
+    A local path is returned as a string.
+    An http URL is read into a buffer and returned as a :class:`BytesIO`.
+
+    When ``encoding`` is not ``utf8`` or ``utf8-lossy``, the whole file is
+    first read in python and decoded using the specified encoding and
+    returned as a :class:`BytesIO` (for usage with ``read_csv``).
+
+    When fsspec is installed, remote file(s) is (are) opened with
+    `fsspec.open(file, **kwargs)` or `fsspec.open_files(file, **kwargs)`.
+    If encoding is not ``utf8`` or ``utf8-lossy``, decoding is handled by
+    fsspec too.
+
+    """
+
+    # Small helper to use a variable as context
+    @contextmanager
+    def managed_file(file: Any) -> Iterator[Any]:
+        try:
+            yield file
+        finally:
+            pass
+
+    has_non_utf8_non_utf8_lossy_encoding = (
+        encoding not in {"utf8", "utf8-lossy"} if encoding else False
+    )
+    encoding_str = encoding if encoding else "utf8"
+
+    if isinstance(file, Path):
+        if has_non_utf8_non_utf8_lossy_encoding:
+            return _check_empty(
+                BytesIO(file.read_bytes().decode(encoding_str).encode("utf8")),
+                context=f"Path ({file!r})",
+            )
+        return managed_file(normalise_filepath(file, False))
+
+    if isinstance(file, str):
+        # make sure that this is before fsspec
+        # as fsspec needs requests to be installed
+        # to read from http
+        if file.startswith("http"):
+            return _process_http_file(file, encoding_str)
+        if _FSSPEC_AVAILABLE:
+            from fsspec.utils import infer_storage_options
+
+            if not has_non_utf8_non_utf8_lossy_encoding:
+                if infer_storage_options(file)["protocol"] == "file":
+                    return managed_file(normalise_filepath(file, False))
+            kwargs["mode"] = "wb"
+            kwargs["encoding"] = encoding
+            return fsspec.open(file, **kwargs)
+
+    if isinstance(file, str):
+        file = normalise_filepath(file, False)
+        if has_non_utf8_non_utf8_lossy_encoding:
+            with open(file, encoding=encoding_str) as f:
```

Comment:
```
polars/io/_utils.py:264:18: PTH123 `open()` should be replaced by `Path.open()`
```

Reply:
@cnpryer in the mergeconflict, I saw that Path(file).open() was added, so I changed to that.